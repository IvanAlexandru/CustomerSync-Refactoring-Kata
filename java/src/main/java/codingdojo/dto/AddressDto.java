package codingdojo.dto;

import lombok.AllArgsConstructor;
import lombok.Data;

/**
 * Address data transfer object
 */
@Data
@AllArgsConstructor
public class AddressDto {
    private String street;
    private String city;
    private String postalCode;
}
