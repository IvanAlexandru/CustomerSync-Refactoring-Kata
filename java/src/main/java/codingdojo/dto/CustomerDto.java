package codingdojo.dto;

import lombok.Data;
import org.hibernate.validator.constraints.NotBlank;

import java.util.List;

/**
 * Customer data transfer object
 */
@Data
public class CustomerDto {
    private AddressDto address;
    @NotBlank
    private String name;
    private String preferredStore;
    private List<ShoppingListDto> shoppingList;
    @NotBlank
    private String externalId;
    private String companyNumber;
    private int bonusPoints;

    public boolean isCompany() {
        return companyNumber != null;
    }
}
