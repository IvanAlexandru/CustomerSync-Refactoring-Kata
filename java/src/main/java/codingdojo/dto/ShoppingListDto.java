package codingdojo.dto;

import lombok.Data;

import java.util.Arrays;
import java.util.List;

/**
 * ShoppingList data transfer object
 */
@Data
public class ShoppingListDto {
    private final List<String> products;

    public ShoppingListDto(String... products) {
        this.products = Arrays.asList(products);
    }
}
