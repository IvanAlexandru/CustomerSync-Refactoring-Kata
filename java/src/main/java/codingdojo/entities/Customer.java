package codingdojo.entities;

import lombok.Data;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

/**
 * Customer database entity
 */
@Data
public class Customer {
    private String id;
    private String externalId;
    private String masterExternalId;
    private Address address;
    private String preferredStore;
    private List<ShoppingList> shoppingLists = new ArrayList<>();
    private String name;
    private CustomerType customerType;
    private String companyNumber;
    private int bonusPoints;

    /**
     * Add shopping list to already existing list of shopping lists
     * @param customerShoppingList - list to be added
     */
    public void addShoppingList(ShoppingList customerShoppingList) {
        shoppingLists.add(customerShoppingList);
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (!(o instanceof Customer)) return false;
        Customer customer = (Customer) o;
        return Objects.equals(externalId, customer.externalId) &&
                Objects.equals(masterExternalId, customer.masterExternalId) &&
                Objects.equals(companyNumber, customer.companyNumber);
    }

    @Override
    public int hashCode() {
        return Objects.hash(externalId, masterExternalId, companyNumber);
    }
}
