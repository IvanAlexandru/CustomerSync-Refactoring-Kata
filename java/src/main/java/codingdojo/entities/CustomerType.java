package codingdojo.entities;

/**
 * Enum defining the possible customer types
 * Values: PERSON, COMPANY
 */
public enum CustomerType {
    PERSON, COMPANY
}
