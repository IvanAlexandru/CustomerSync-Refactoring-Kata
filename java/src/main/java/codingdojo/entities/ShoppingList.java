package codingdojo.entities;

import lombok.Data;

import java.util.List;

/**
 * ShoppingList database entity
 */
@Data
public class ShoppingList {
    private final List<String> products;

    public ShoppingList(List<String> products) {
        this.products = products;
    }
}
