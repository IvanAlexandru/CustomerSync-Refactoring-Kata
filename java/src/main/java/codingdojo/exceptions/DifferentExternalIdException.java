package codingdojo.exceptions;

/**
 * Occurs when externalId of customerDto is different than the one of the customer entity
 */
public class DifferentExternalIdException extends RuntimeException {
    public DifferentExternalIdException(String s) {
        super(s);
    }
}
