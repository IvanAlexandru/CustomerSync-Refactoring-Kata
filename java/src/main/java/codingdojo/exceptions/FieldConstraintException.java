package codingdojo.exceptions;

/**
 * Occurs when {@link codingdojo.dto.CustomerDto} fails field validation
 */
public class FieldConstraintException extends RuntimeException {
    public FieldConstraintException(String s) {
        super(s);
    }
}
