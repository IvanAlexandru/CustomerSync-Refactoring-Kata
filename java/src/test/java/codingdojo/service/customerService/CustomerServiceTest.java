package codingdojo.service.customerService;

import codingdojo.dao.FakeDatabase;
import codingdojo.entities.CustomerType;
import codingdojo.dto.AddressDto;
import codingdojo.dto.CustomerDto;
import codingdojo.dto.ShoppingListDto;
import codingdojo.entities.Customer;
import codingdojo.entities.ShoppingList;
import codingdojo.exceptions.DifferentExternalIdException;
import codingdojo.exceptions.FieldConstraintException;
import codingdojo.exceptions.WrongCustomerTypeException;
import codingdojo.printers.ExternalCustomerPrinter;
import codingdojo.service.CustomerService;
import org.approvaltests.Approvals;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;

import java.util.Arrays;

import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertTrue;


public class CustomerServiceTest {

    @Test
    public void syncCompanyByExternalId(){
        String externalId = "12345";

        CustomerDto customerDto = createExternalCompany();
        customerDto.setExternalId(externalId);

        Customer customer = createCustomerWithSameCompanyAs(customerDto);
        customer.setExternalId(externalId);

        FakeDatabase db = new FakeDatabase();
        db.addCustomer(customer);
        CustomerService customerService = new CustomerService(db);

        StringBuilder toAssert = printBeforeState(customerDto, db);

        // ACT
        boolean created = customerService.syncWithDataLayer(customerDto);

        assertFalse(created);
        printAfterState(db, toAssert);
        Approvals.verify(toAssert);
    }

    @Test
    public void syncPrivatePersonByExternalId(){
        String externalId = "12345";

        CustomerDto customerDto = createExternalPrivatePerson();
        customerDto.setExternalId(externalId);

        Customer customer = new Customer();
        customer.setCustomerType(CustomerType.PERSON);
        customer.setId("67576");
        customer.setExternalId(externalId);

        FakeDatabase db = new FakeDatabase();
        db.addCustomer(customer);
        CustomerService customerService = new CustomerService(db);

        StringBuilder toAssert = printBeforeState(customerDto, db);

        // ACT
        boolean created = customerService.syncWithDataLayer(customerDto);

        assertFalse(created);
        printAfterState(db, toAssert);
        Approvals.verify(toAssert);
    }

    @Test
    public void syncShoppingLists(){
        String externalId = "12345";

        CustomerDto customerDto = createExternalCompany();
        customerDto.setExternalId(externalId);

        Customer customer = createCustomerWithSameCompanyAs(customerDto);
        customer.setExternalId(externalId);
        customer.addShoppingList(new ShoppingList(Arrays.asList("eyeliner", "blusher")));

        FakeDatabase db = new FakeDatabase();
        db.addCustomer(customer);
        CustomerService customerService = new CustomerService(db);

        StringBuilder toAssert = printBeforeState(customerDto, db);

        // ACT
        boolean created = customerService.syncWithDataLayer(customerDto);

        assertFalse(created);
        printAfterState(db, toAssert);
        Approvals.verify(toAssert);
    }

    @Test
    public void syncNewCompanyCustomer(){

        CustomerDto customerDto = createExternalCompany();
        customerDto.setExternalId("12345");

        FakeDatabase db = new FakeDatabase();
        CustomerService customerService = new CustomerService(db);

        StringBuilder toAssert = printBeforeState(customerDto, db);

        // ACT
        boolean created = customerService.syncWithDataLayer(customerDto);

        assertTrue(created);
        printAfterState(db, toAssert);
        Approvals.verify(toAssert);
    }

    @Test
    public void syncNewPrivateCustomer(){

        CustomerDto customerDto = createExternalPrivatePerson();
        customerDto.setExternalId("12345");

        FakeDatabase db = new FakeDatabase();
        CustomerService customerService = new CustomerService(db);

        StringBuilder toAssert = printBeforeState(customerDto, db);

        // ACT
        boolean created = customerService.syncWithDataLayer(customerDto);

        assertTrue(created);
        printAfterState(db, toAssert);
        Approvals.verify(toAssert);
    }

    @Test
    public void wrongCustomerTypeExceptionWhenExistingCustomerIsPerson() {
        String externalId = "12345";

        CustomerDto customerDto = createExternalCompany();
        customerDto.setExternalId(externalId);

        Customer customer = new Customer();
        customer.setCustomerType(CustomerType.PERSON);
        customer.setId("45435");
        customer.setExternalId(externalId);

        FakeDatabase db = new FakeDatabase();
        db.addCustomer(customer);
        CustomerService customerService = new CustomerService(db);

        StringBuilder toAssert = printBeforeState(customerDto, db);

        Assertions.assertThrows(WrongCustomerTypeException.class, () -> {
            customerService.syncWithDataLayer(customerDto);
        }, printAfterState(db, toAssert).toString());

        Approvals.verify(toAssert);
    }

    @Test
    public void syncByExternalIdButCompanyNumbersConflict(){
        String externalId = "12345";

        CustomerDto customerDto = createExternalCompany();
        customerDto.setExternalId(externalId);

        Customer customer = createCustomerWithSameCompanyAs(customerDto);
        customer.setExternalId(externalId);
        customer.setCompanyNumber("000-3234");

        FakeDatabase db = new FakeDatabase();
        db.addCustomer(customer);
        CustomerService customerService = new CustomerService(db);

        StringBuilder toAssert = printBeforeState(customerDto, db);

        // ACT
        boolean created = customerService.syncWithDataLayer(customerDto);

        assertTrue(created);
        printAfterState(db, toAssert);
        Approvals.verify(toAssert);
    }


    @Test
    public void syncByCompanyNumber(){
        String companyNumber = "12345";

        CustomerDto customerDto = createExternalCompany();
        customerDto.setCompanyNumber(companyNumber);

        Customer customer = createCustomerWithSameCompanyAs(customerDto);
        customer.setCompanyNumber(companyNumber);
        customer.addShoppingList(new ShoppingList(Arrays.asList("eyeliner", "mascara", "blue bombe eyeshadow")));

        FakeDatabase db = new FakeDatabase();
        db.addCustomer(customer);
        CustomerService customerService = new CustomerService(db);

        StringBuilder toAssert = printBeforeState(customerDto, db);

        // ACT
        boolean created = customerService.syncWithDataLayer(customerDto);

        assertFalse(created);
        printAfterState(db, toAssert);
        Approvals.verify(toAssert);
    }

    @Test
    public void syncByCompanyNumberWithConflictingExternalId(){
        String companyNumber = "12345";

        CustomerDto customerDto = createExternalCompany();
        customerDto.setCompanyNumber(companyNumber);
        customerDto.setExternalId("45646");

        Customer customer = createCustomerWithSameCompanyAs(customerDto);
        customer.setCompanyNumber(companyNumber);
        customer.setExternalId("conflicting id");

        FakeDatabase db = new FakeDatabase();
        db.addCustomer(customer);
        CustomerService customerService = new CustomerService(db);

        StringBuilder toAssert = printBeforeState(customerDto, db);

        // ACT
        Assertions.assertThrows(DifferentExternalIdException.class, () -> {
            customerService.syncWithDataLayer(customerDto);
        }, printAfterState(db, toAssert).toString());

        Approvals.verify(toAssert);
    }

    @Test
    public void wrongCustomerTypeExceptionWhenExistingCustomerIsCompany() {
        String externalId = "12345";

        CustomerDto customerDto = createExternalPrivatePerson();
        customerDto.setExternalId(externalId);

        Customer customer = new Customer();
        customer.setCustomerType(CustomerType.COMPANY);
        customer.setCompanyNumber("32423-342");
        customer.setId("45435");
        customer.setExternalId(externalId);

        FakeDatabase db = new FakeDatabase();
        db.addCustomer(customer);
        CustomerService customerService = new CustomerService(db);

        StringBuilder toAssert = printBeforeState(customerDto, db);

        Assertions.assertThrows(WrongCustomerTypeException.class, () -> {
            customerService.syncWithDataLayer(customerDto);
        }, printAfterState(db, toAssert).toString());

        Approvals.verify(toAssert);
    }

    @Test
    public void syncCompanyByExternalIdWithNonMatchingMasterId(){
        String externalId = "12345";

        CustomerDto customerDto = createExternalCompany();
        customerDto.setExternalId(externalId);

        Customer customer = createCustomerWithSameCompanyAs(customerDto);
        customer.setExternalId(externalId);
        customer.setName("company 1");

        Customer customer2 = new Customer();
        customer2.setCompanyNumber(customerDto.getCompanyNumber());
        customer2.setCustomerType(CustomerType.COMPANY);
        customer2.setId("45435234");
        customer2.setMasterExternalId(externalId);
        customer2.setName("company 2");

        FakeDatabase db = new FakeDatabase();
        db.addCustomer(customer);
        db.addCustomer(customer2);
        CustomerService customerService = new CustomerService(db);

        StringBuilder toAssert = printBeforeState(customerDto, db);

        // ACT
        boolean created = customerService.syncWithDataLayer(customerDto);

        assertFalse(created);
        printAfterState(db, toAssert);
        Approvals.verify(toAssert);
    }

    @Test
    public void fieldConstraintExceptionName() {
        CustomerDto customerDto = createExternalCompany();
        customerDto.setName("");

        Customer customer = new Customer();
        customer.setCustomerType(CustomerType.PERSON);
        customer.setExternalId("12345");
        customer.setId("45435");

        FakeDatabase db = new FakeDatabase();
        db.addCustomer(customer);
        CustomerService customerService = new CustomerService(db);

        StringBuilder toAssert = printBeforeState(customerDto, db);

        Assertions.assertThrows(FieldConstraintException.class, () -> {
            customerService.syncWithDataLayer(customerDto);
        }, printAfterState(db, toAssert).toString());

        Approvals.verify(toAssert);
    }

    private CustomerDto createExternalPrivatePerson() {
        CustomerDto customerDto = new CustomerDto();
        customerDto.setExternalId("12345");
        customerDto.setName("Joe Bloggs");
        customerDto.setAddress(new AddressDto("123 main st", "Stockholm", "SE-123 45"));
        customerDto.setPreferredStore("Nordstan");
        customerDto.setBonusPoints(50);
        customerDto.setShoppingList(Arrays.asList(new ShoppingListDto("lipstick", "foundation")));
        return customerDto;
    }


    private CustomerDto createExternalCompany() {
        CustomerDto customerDto = new CustomerDto();
        customerDto.setExternalId("12345");
        customerDto.setName("Acme Inc.");
        customerDto.setAddress(new AddressDto("123 main st", "Helsingborg", "SE-123 45"));
        customerDto.setCompanyNumber("470813-8895");
        customerDto.setShoppingList(Arrays.asList(new ShoppingListDto("lipstick", "blusher")));
        return customerDto;
    }

    private Customer createCustomerWithSameCompanyAs(CustomerDto customerDto) {
        Customer customer = new Customer();
        customer.setCompanyNumber(customerDto.getCompanyNumber());
        customer.setCustomerType(CustomerType.COMPANY);
        customer.setId("45435");
        return customer;
    }

    private StringBuilder printBeforeState(CustomerDto customerDto, FakeDatabase db) {
        StringBuilder toAssert = new StringBuilder();
        toAssert.append("BEFORE:\n");
        toAssert.append(db.printContents());

        toAssert.append("\nSYNCING THIS:\n");
        toAssert.append(ExternalCustomerPrinter.print(customerDto, ""));
        return toAssert;
    }

    private StringBuilder printAfterState(FakeDatabase db, StringBuilder toAssert) {
        toAssert.append("\nAFTER:\n");
        toAssert.append(db.printContents());
        return toAssert;
    }
}
