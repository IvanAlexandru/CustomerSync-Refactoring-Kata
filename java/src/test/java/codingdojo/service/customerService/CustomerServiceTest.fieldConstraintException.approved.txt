BEFORE:
Fake Database.
All Customers {

    Customer {
        externalId='12345'
        masterExternalId='null'
        companyNumber='null'
        internalId='45435'
        name='null'
        customerType=PERSON
        preferredStore='null'
        bonusPoints='0'
        address='null'
        shoppingLists=[]
    }

}
All Shopping Lists
[]
SYNCING THIS:
ExternalCustomer {
    externalId=''
    companyNumber='470813-8895'
    name=''
    preferredStore='null'
    bonusPoints='0'
    address='123 main st, SE-123 45 Helsingborg'
    shoppingLists=[[lipstick, blusher]]
}
AFTER:
Fake Database.
All Customers {

    Customer {
        externalId='12345'
        masterExternalId='null'
        companyNumber='null'
        internalId='45435'
        name='null'
        customerType=PERSON
        preferredStore='null'
        bonusPoints='0'
        address='null'
        shoppingLists=[]
    }

}
All Shopping Lists
[]